cargo-screeps/docs/initialization-header.md

```diff
@@ -0,0 +1,136 @@
+Writing an initialization header
+================================
+
+The initialization header is a piece of code controlling how the WASM environment starts up. As
+screeps is a resource constrained environment, and your rust code may want to interop with other
+code, this initialization is completely controllable.
+
+The default initialization header is as follows:
+
+```js
+"use strict";
+// Create a global variable to store the wasm module. This is the compiled uninstantiated code.
+// If `new WebAssembly.Instance` times out, this will helpfully remain here to be reused next tick.
+let wasm_module = null;
+
+// This function does the initialization. It's separate from `module.exports.loop` so JS code can
+// run `module.exports.loop = wasm_initialize` to reset the WASM vm next tick.
+function wasm_initialize() {
+    // If we haven't compiled the code, let's do that.
+    if (wasm_module == null) {
+        // Fetch the source byte array. `wasm_fetch_module_bytes` just returns
+        // require('wasm_module_name'), but it's useful to have as a separate function
+        // so we don't hardcode the module name.
+        let wasm_bytes = wasm_fetch_module_bytes();
+        // Compile the module. I've found that this can time out, but because of however it works,
+        // `wasm_module` will still be initialized.
+        wasm_module = new WebAssembly.Module(wasm_bytes);
+    }
+    // Initialize `stdweb`. This returns an object with an `initialize` function and `imports`
+    // object, meant to be used as bellow:
+    let stdweb_vars = wasm_create_stdweb_vars();
+    // Create the WebAssembly instance.
+    let wasm_instance = new WebAssembly.Instance(wasm_module, stdweb_vars.imports);
+    // Run the remaining stdweb initialization. This is fairly short, just some things to make sure
+    // callbacks work right and such. This will also run the rust `main` function.
+    stdweb_vars.initialize(wasm_instance);
+    // Now we run the actual main loop - this assumes that the rust `main` function overwrites
+    // `module.exports.loop`.
+    module.exports.loop();
```

Comment:
Isn't that equivalent to recursively calling `wasm_initialize`? Or does that bind beforehand and now it basically injects `wasm_initialize` before `module.exports.loop()` is called?

Reply:
The second, if I understand correctly.

`stdweb_vars.initialize` will call the `fn main()` in rust code, and it's assumed that that function does `module.exports.loop = @{some_rust_fn};`.

So- yes, this is essentially injecting `wasm_initialize` before the regular `loop` function.